INT-3989: Add `FileReadingMessageSource.WatchServiceDirectoryScanner`

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3989,
https://jira.spring.io/browse/INT-3990,
https://jira.spring.io/browse/INT-3988
- Deprecate top-level `WatchServiceDirectoryScanner` because of inconsistency around `Lifecycle` and shared `directory` property
- Copy/paste its logic into the `FileReadingMessageSource.WatchServiceDirectoryScanner` to hide that inconsistency, but still get a gain from the `WatchService` benefits
- Add support for the `StandardWatchEventKinds.ENTRY_MODIFY` and `StandardWatchEventKinds.ENTRY_DELETE` events in the `FileReadingMessageSource.WatchServiceDirectoryScanner`
- Introduce `useWatchService` option to switch to the internal `FileReadingMessageSource.WatchServiceDirectoryScanner`
- Make `CompositeFileListFilter` also as `ResettableFileListFilter`
- Deprecate weird `FileReadingMessageSource.onSend()` method and remove its usage from tests
- Modify `WatchServiceDirectoryScannerTests` for the new logic
- Document changes

artembilan committed:
INT-3989: Add `FileReadingMessageSource.WatchServiceDirectoryScanner`

JIRA: https://jira.spring.io/browse/INT-3989,
https://jira.spring.io/browse/INT-3990,
https://jira.spring.io/browse/INT-3988

* Deprecate top-level `WatchServiceDirectoryScanner` because of inconsistency around `Lifecycle` and shared `directory` property
* Copy/paste its logic into the `FileReadingMessageSource.WatchServiceDirectoryScanner` to hide that inconsistency, but still get a gain from the `WatchService` benefits
* Add support for the `StandardWatchEventKinds.ENTRY_MODIFY` and `StandardWatchEventKinds.ENTRY_DELETE` events in the `FileReadingMessageSource.WatchServiceDirectoryScanner`
* Introduce `useWatchService` option to switch to the internal `FileReadingMessageSource.WatchServiceDirectoryScanner`
* Make `CompositeFileListFilter` also as `ResettableFileListFilter`
* Deprecate weird `FileReadingMessageSource.onSend()` method and remove its usage from tests
* Modify `WatchServiceDirectoryScannerTests` for the new logic
* Document changes

artembilan committed:
Add `MODIFY` and `DELETE` test coverage

Optimize the `filesFromEvents()` logic replacing item with the fresh event sources.
Remove the item from the result set in case of `DELETE` event, because file removal generates both `MODIFY` and `DELETE` events.

artembilan commented:
Pushed some optimization and testing.

artembilan committed:
* Add `FileReadingMessageSource.setWatchEvents` to allow to listen to the specific events,
not only `CREATE` or all of them.
* Modify tests and docs to reflect the new API

artembilan commented:
Pushed the `watch-events` support.

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileReadingMessageSource.java:
```diff
@@ -236,12 +257,57 @@ public void setScanEachPoll(boolean scanEachPoll) {
+	 */
+	public void setWatchEvents(WatchEventType... watchEvents) {
+		Assert.notEmpty(watchEvents, "'watchEvents' must not be empty.");
+		this.watchEvents = watchEvents;
```
`Arrays.copyOf(watchEvents, watchEvents.length)`

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileReadingMessageSource.java:
```diff
@@ -236,12 +257,57 @@ public void setScanEachPoll(boolean scanEachPoll) {
+	 * @see #setUseWatchService
+	 */
+	public void setWatchEvents(WatchEventType... watchEvents) {
+		Assert.notEmpty(watchEvents, "'watchEvents' must not be empty.");
```
`Assert.noNullElements()`
`Assert.start(!this.running, "Cannot change watch events while running")`

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileReadingMessageSource.java:
```diff
@@ -87,10 +102,16 @@
 
+	private boolean useWatchService;
+
+	private WatchEventType[] watchEvents = new WatchEventType[] { WatchEventType.CREATE };
```
Perhaps we should support create + delete by default - (or even all 3) - consider add + remove between polls - we'd emit a message for a file that no longer exists.

artembilan committed:
Address PR comments

* Improve `FileReadingMessageSource.setWatchEvents()`
* Add `file.exists()` before adding file from event.

artembilan commented:
Pushed.
> consider add + remove between polls

See commit comments.

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/FileReadingMessageSource.java:
```diff
@@ -495,12 +499,14 @@ public boolean isRunning() {
+									files.remove(file);
+									files.add(file);
+								}
 							}
```
We should probably add a debug log here "File removed since watch event received".

artembilan replied:
Indeed! :+1:

artembilan committed:
Add more DEBUG logs into the `WatchServiceDirectoryScanner`

With the fact of those logs provide more optimizations:
* Don't register the same directory for watching:
  - use the `ConcurrentMap<Path, WatchKey> pathKeys` to track registrations
  - any modification within the directory causes the `ENTRY_MODIFY` for the directory as well.
    So, skip such an event exactly for the directory during `walkDirectory`
* Add debug logs in case of `ENTRY_DELETE`

artembilan commented:
Pushed.

garyrussell commented:
Merged as 7b1f77ac1ffd6dc074cea9a4f48cb3c03343d8bc
